### PR TITLE
Add test for server-side/isomorphic rendering

### LINKED ISSUE #370: Add tests for SSR to prevent regressions
Since the plan is to ship SSR/isomorphic rendering support with v1.4.0, there should probably be some minimal functional tests to prevent regressions as users begin to depend on this feature.

Reference: #360, [wiki](https://github.com/palantir/blueprint/wiki/Frequently-asked-questions#5-does-blueprint-support-server-side-rendering) cc/ @adidahiya @codebling (nice work dude)

### PR DESCRIPTION
**Pull request by codebling**
#### Changes proposed in this pull request:

- Add regression for server-side/isomorphic rendering added in #360 (test that all Components can be rendered server-side)

See list of potential issues in comments below.

### CODE REVIEW

**Comment by codebling** [comment]
Potential issues with this PR:
  * The test is written in JavaScript, not TypeScript. This is because stock Mocha (unlike Karma) does not have a TS pre-processor, not can it be be piped into. We can use TypeScript if we're willing to have a temp/output directory in the project. If that's the case, let me know where this directory should be located and I can make the changes. 
  * Commit [`db59e23`](https://github.com/palantir/blueprint/pull/381/commits/db59e23a0fcec2956310273dbc8239895ea83387) is not strictly necessary for this PR (although it improves the modularisation of the code). It can be reverted within the scope of this PR and added separately, if needed. 
  * The test uses the shallow renderer from `ReactTestUtils`. This is not in keeping with the rest of the project, where we use the shallow renderer from `enzyme`. This was done because for certain Components, we need to pass specific Props and Children, and I wasn't able to get this to work with Enzyme. If someone is able to figure out how to get this to work, we can use the Enzyme shallow renderer.

**Comment by codebling** [comment]
I'm not sure why CircleCI won't trigger. I've authorised the CircleCI application, re-added the deploy key and I'm following the project. I've tried adding CircleCI to the project as an integration and pushing commits, it still won't trigger.

File: Gulpfile.js [hunk 1]
```diff
@@ -107,6 +112,7 @@ const projects = [
         cwd: "packages/table/",
         dependencies: ["core"],
         copy: false,
+        isotest: true,
```
Comment (giladgray) [inline Gulpfile.js]: seems like isotest and karma are 1:1 -- if a package has karma tests then it'll also need isotests. so we probably don't need both tags. could rename them all to simply "test" ?

File: gulp/util/create-config.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2015 Palantir Technologies, Inc. All rights reserved.
+ */
+"use strict";
+
+module.exports = function createConfig (project) {
```
Comment (giladgray) [inline gulp/util/create-config.js]: at the very least, filename should have some mention of "karma" so we know what it's for. like its sibling, `webpack-config.js`.

but i had no problem with this config living in karma.js directly as it's only needed there.

Comment (giladgray) [inline gulp/util/create-config.js]: plus now i can't tell what you changed cuz it's all green.

File: package.json [hunk 3]
```diff
@@ -42,6 +42,7 @@
     "gulp-insert": "0.5.0",
     "gulp-less": "3.1.0",
     "gulp-load-plugins": "1.3.0",
+    "gulp-mocha": "^3.0.1",
```
Comment (giladgray) [inline package.json]: exact version please

File: packages/core/test.iso/index.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (giladgray) [inline packages/core/test.iso/index.js]: there's a root `/test/` directory that would be a great place to put the repeated parts of this file. then these `test.iso/index.js` files would import that shared code and pass their own list of components and options.

File: Gulpfile.js (hunk 1 above)
Comment (codebling) [inline Gulpfile.js]: The mapping is indeed 1:1. If it looks like it will stay that way, we can merge them. `isotest` has a compile/build dependency which `karma` does not. If we combine the tasks, the combined task will also have this dependency. If you can live with this, I think the merged task will be an improvement. (Def think `karma` and `isotest` are not the best names)

Comment (giladgray) [inline Gulpfile.js]: i'm suggesting combining the _config flags_, not the test tasks. 
definitely should be separate gulp tasks.

File: gulp/util/create-config.js (hunk 2 above)
Comment (codebling) [inline gulp/util/create-config.js]: I am a micro-commit maniac, so you check [db59e23](https://github.com/palantir/blueprint/pull/381/commits/db59e23a0fcec2956310273dbc8239895ea83387). It strictly refactors the `createConfig` function out of that file. 

I thought that code would need to be shared but in the end it wasn't. This change is not needed for this PR, I almost took it up but kept it in as a last-minute decision because it cleaned up `karma.js` significantly. It should definitely be named `karma-config.js`. 

Can add a revert commit to the PR if you'd like to keep the code within `karma.js`. Just give me your final call on revert vs rename.

File: packages/core/test.iso/index.js (hunk 4 above)
Comment (codebling) [inline packages/core/test.iso/index.js]: That's a great idea

File: gulp/util/create-config.js (hunk 2 above)
Comment (giladgray) [inline gulp/util/create-config.js]: awesome! this looks good and cleans the file up nicely. thanks for explaining 👍

File: Gulpfile.js (hunk 1 above)
Comment (codebling) [inline Gulpfile.js]: That's understood, but aren't the *config flags* used to auto-generate the tasks for each package?  I'll have to look it again. The build system is non-trivial

File: gulp/util/create-config.js (hunk 2 above)
Comment (codebling) [inline gulp/util/create-config.js]: Perfect. I'll rename the file.

File: package.json (hunk 3 above)
Comment (codebling) [inline package.json]: Will do. Consider [`npm shrinkwrap`](https://docs.npmjs.com/cli/shrinkwrap) if the goal is freezing the dependency tree. Using exact numbers brings a maintenance cost and still does not lock down 2nd level dependencies.

File: Gulpfile.js (hunk 1 above)
Comment (giladgray) [inline Gulpfile.js]: yeah that's why i suggested changing to `tests` so we'd have `test-karma-core` and `test-iso-core`.

Comment (giladgray) [inline Gulpfile.js]: this change may not be worth it--it's just easier and more flexible to have separate flags.

File: package.json (hunk 3 above)
Comment (giladgray) [inline package.json]: heck no we deleted our shrinkwrap a while back and haven't looked back. in my experience, npm shrinkwrap only causes headaches.

we'll move to yarn as soon as we can.

File: gulp/isotest.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright 2015 Palantir Technologies, Inc. All rights reserved.
+ */
+"use strict";
+
+module.exports = (gulp, plugins, blueprint) => {
+    const rs = require("run-sequence").use(gulp);
+    const mocha = require("gulp-mocha");
+
+    blueprint.projectsWithBlock("isotest").forEach((project) => {
+        gulp.task(`isotest-${project.id}`, [`typescript-compile-${project.id}`], (done) => {
```
Comment (giladgray) [inline gulp/isotest.js]: please remove the typescript dependency here. it's going to end up compiling each project multiple times throughout the `build` flow. instead, add a single dependency on `typescript-compile` below when you define the `isotest` wrapper.

or even better, don't add any dependencies and just assume that we know to compile before running tests (that's a safe assumption)

case in point: at this point in the circle build we've already run `gulp build` and now you're asking to _re-run_ the most time-consuming parts of that build (`tsc`)

File: Gulpfile.js (hunk 1 above)
Comment (codebling) [inline Gulpfile.js]: I'll have another look and see what can be done

File: gulp/isotest.js (hunk 5 above)
Comment (codebling) [inline gulp/isotest.js]: This was not lost on me, but creating a task which will fail or give false results if you haven't compiled recently seems even worse. How about this: I create a second group of `isotest` task (perhaps `isotest-nobuild`?) which don't depend on the `typescript-compile` tasks. We use the 'nobuild' variant with CircleCI. The 'nobuild' variant can be run as easily as the 'buid' one, but the explicit naming makes the consequences clear, at least. 

For what it's worth, this is happening because `run-sequence` doesn't promote the dependencies to Gulp. If I had more time/motivation, `run-sequence` could be patched or re-written to correctly promote the dependencies, at which point Gulp would not re-run the tasks which had already been executed. Alternatively, the `typescript-compile`/ build tasks could perhaps be modified to test for freshness, though this seems rather hacky and suboptimal to me. Running tasks in sequence will be much easier when [Gulp 4 is released](https://github.com/gulpjs/gulp/milestone/1), but given that it's already 2 years in the works, who knows if it will ever be released.

Comment (codebling) [inline gulp/isotest.js]: @giladgray  what do you think about that solution?

Comment (giladgray) [inline gulp/isotest.js]: @codebling if you use `blueprint.task()` you'll get the no-dependency behavior for free, though I wouldn't presume to make you figure out how that thing works. in short, it would generate `isotest-mocha-core` and `isotest-mocha-w-core`; the `-w-` task (short for "watch") runs without any dependencies.

File: packages/core/test.iso/index.js (hunk 4 above)
Comment (giladgray) [inline packages/core/test.iso/index.js]: @codebling can you please make this refactor?

File: gulp/aliases.js [hunk 6]
```diff
@@ -26,7 +26,7 @@ module.exports = (gulp) => {
     gulp.task("build", (done) => rs("clean", "compile", "webpack-compile-docs", done));
 
     // build code, run unit tests, terminate
-    gulp.task("test", ["karma"]);
+    gulp.task("test", ["karma", "isotest"]);
```
Comment (giladgray) [inline gulp/aliases.js]: merge conflict here as i added `test-dist` yesterday

File: circle.yml [hunk 7]
```diff
@@ -31,7 +31,8 @@ dependencies:
 test:
   override:
     - gulp check
-    - gulp test
+    - gulp karma
+    - gulp isotest
```
Comment (giladgray) [inline circle.yml]: i just added `test-dist` task yesterday so either add that here or revert this change back to `gulp test`

File: gulp/isotest.js (hunk 5 above)
Comment (giladgray) [inline gulp/isotest.js]: look at how the `typescript-compile` tasks are defined in typescript.js. you'll want to do something similar here. depending on `["typescript-compile-*"]` will auto-expand the * to corresponding project id.

**Comment by giladgray** [comment]
@codebling i resolved merge conflicts (through github UI, very cool 😎 ). going to merge this into master because the best parts are ready and i'll do a refactor pass today. thank you 🎩 !

**Comment by codebling** [comment]
@giladgray great. Still haven't had a chance to check the changes in 595, let me know if you need anything from me.
